client/structs/broadcaster.go

```diff
@@ -111,6 +135,11 @@ func (b *AllocBroadcaster) Listen() *AllocListener {
 
 	ch := make(chan *structs.Allocation, listenerCap)
 
+	// Send last update if there was one
+	if b.last != nil {
+		ch <- b.last
```

Comment:
This also fixes the deadlock by making sure all listeners receive the latest version of the alloc even if the alloc has terminated and broadcaster closed.

This means the deadlock is fixed *2 ways* which I'm not always a fan of. I think the `Terminated()` check in the watcher should be sufficient to fix it, but I think a bit of defensiveness is warranted in this circumstance. It makes the code less brittle and the broadcaster easier to use. Without this everywhere we use the broadcaster would need to make its own Terminated() check to prevent a similar deadlock.

Reply:
This approach sounds good.

I personally like this listener getting latest approach more from a design prospective.  It makes sense for new listeners to get latest value first, followed by new updates.  The alternative typically is to look up value first than listen to updates, introducing a race condition or requiring complex watch offset semantics - so would be +1 for keeping it as-is.

I would add a comment about the ordering of this vs channel closed - and I see you already added a test for it :).

Reply:
Great points. Added a comment because as you can probably guess I got the ordering wrong the first try!